### DISCUSSION TITLE
I'll be taking over core development of vector

### DISCUSSION THREAD

**Discussion by jpivarski** [scikit-hep/vector#30] [General]
Hi watchers and @nsmith-, @jrueb, @marinang, and anyone else who has been interested in vectorized vectors!

A little over a month ago, @henryiii and I worked out a plan in which he would become in charge of the scikit-hep/awkward-1.0 build and deployment and I would bring the scikit-hep/vector package up to 1.0 status. I said I'd start that in March, and managed to get to it in time, as it's now March 2. I think this will take about a month (and then Uproot 4 file-writing is probably next).

The default branch for this project has already been set to `develop`, rather than `main`. I'll be doing this development in PRs that target `develop`.

I'm going to start by clearing out `src/vector`, `tests`, and the documentation files in `docs`, leaving the structure of everything else as it is: package configuration, continuous integration, the structure of the documentation, etc. This is to "get my head around" what is done and what isn't done—I'll be looking at the old code, and the result will have largely the same structure. But, for instance, I'll be turning on the type safety of the protocols (which ensure that we haven't missed any methods) at the end of development, rather than at the beginning.

One structural change that I do plan to make is to make the NumPy implementation a subclass of `np.ndarray` again (as it was in [hepvector](https://github.com/henryiii/hepvector)). That forces the components to be contiguous in memory, which is a sometimes-desirable feature. Awkward Array handles the case when it isn't desired.

I'll be looking closely at all of the other Lorentz vector projects for structure and naming conventions:

   * The `main` branch of this repo and [henryiii/hepvector](https://github.com/henryiii/hepvector).
   * ROOT's [TLorentzVector](https://root.cern.ch/doc/master/classTLorentzVector.html) and especially [Math::LorentzVector](https://root.cern.ch/doc/master/classROOT_1_1Math_1_1LorentzVector.html), as the latter handles multiple coordinate systems as we want here.
   * [scikit-hep/math](https://github.com/scikit-hep/scikit-hep/tree/master/skhep/math), where we specifically thought through what the Scikit-HEP Lorentz vector interface should look like (3 years ago). That's closely related to the [uproot-methods TLorentzVector class](https://github.com/scikit-hep/uproot3-methods/blob/master/uproot3_methods/classes/TLorentzVector.py)—one was made to mirror the other (though I can't remember which).
   * [coffea.nanoevents.methods.vector](https://coffeateam.github.io/coffea/modules/coffea.nanoevents.methods.vector.html), as we want to make the transition from Coffea to vector as seamless as possible.
   * An email I've been keeping from @jrueb (sent to cms-coffea) about "Extending Nanoevents vector".

I'm mostly reiterating the goals of Vector as @henryiii has described them, but it has 8 principles:

   1. 2D, 3D, and Lorentz, are distinct (non-inheriting) classes.
   2. Arrays of vectors (arrays with behavioral methods) are distinct from individual vectors (simple Python objects), but they have all the same methods and behave the same ways.
   3. Coordinate systems determine the _member data_ that each class has: Cartesian, cylindrical, spherical, etc. These classes should be carefully arranged so that we can mix Cartesian vs polar azimuthal with _η_ vs _θ_ vs _cos(θ)_ vs _cot(θ)_ longitudinal with _E_ vs _m_ temporal.
   4. Different storage/computation backends for arrays of vectors: NumPy, Awkward, and possibly CuPy, TensorFlow, Torch, and JAX.
   5. Separate the mathematical functions from the backends, since the same Python expressions mostly duck-type everywhere. (The first argument of all of these functions should be `lib`, so that `numpy` can be replaced with `cupy` or `jax.numpy`, etc.)
   6. Numba extensions for NumPy and Awkward, which also implies a lowered type for individual vectors.
   7. Base arrays/objects have purely geometric names for their arguments, such as `perp` instead of `pt`, but overloads mix in synonyms.
   8. Another mix-in keeps track of affine (position) vs vector (direction), mostly for strong-typing these concepts. For instance, affine `-` affine → vector, but affine `+` affine is an error, while all operations are available to vectors. (This is like ROOT's [Math::PositionVector3D](https://root.cern.ch/doc/master/classROOT_1_1Math_1_1PositionVector3D.html) versus [Math::DisplacementVector3D](https://root.cern.ch/doc/master/classROOT_1_1Math_1_1DisplacementVector3D.html).)

The order of development will be roughly as follows:

   1. Decide on a complete set of coordinate systems and name them. This will set a naming convention and scope for the mathematical functions.
   2. Implement the mathematical functions, driven by tests against ROOT's Math::LorentzVector (through PyROOT). The functions can be tested before any classes exist by calling them directly in the test suite (unit, rather than integration, tests).
   3. Develop the class hierarchy for NumPy only. This involves the inheritance structure of the momentum-name mixins and the affine-vs-vector mixins.
   4. Extend that to CuPy as an easy example of a different library with a different `lib` to pass into all the math functions.
   5. Develop the Awkward behaviors, which are not strictly methods on a class (some are through `ak.behavior`).
   6. Lower the NumPy backend to Numba, which involves developing a lowered "individual vector" type.
   7. Lower the Awkward backend to Numba, too, which also involves `ak.behavior`.
   8. Maybe do JAX, too, because I've had some experience with that recently and differentiable Lorentz methods would be cool.
   9. Type-annotate everything in one pass, just as I've found it useful to write all the docstrings in one pass.
   10. Write all the docstrings in one pass, too.
   11. Reinstate the protocols to ensure that no methods are missing or incorrectly typed.
   12. Generate reference documentation from the docstrings (i.e. write .rst shells).

By this point, we'd want to release some versions and get some feedback from users. `develop` should be merged into `main` and releases/release candidates should be deployed. Hopefully, I can get to step 12 by the end of the month, but we'll see.

Last minute thoughts that I'll be thinking about tonight: do we need Transformations to be objects? A Transformation with both a translation and rotation would act on affines differently from vectors, and they would also be parameterized by coordinate system. (The vectors in ROOT's `Math` library have this.) Then would there be an array of Transformations to go along with an individual Transformation, like arrays? I'll have to think about whether that makes things easier or harder.

**Comment by nsmith-** [comment]
Re: transforms as objects, Lindsey floated [boost tensors](https://github.com/CoffeaTeam/coffea/pull/447#discussion_r574810303) in coffea once so maybe worth having.

**Reply by jpivarski** [reply]
I read that and it helped firm my opinion.

My compromise between generality and the amount of work to do is this: Transformations should be first-class objects like Vectors, but Transformations don't have to exist for all coordinate systems (which would create an _n²_ problem or worse with all the pairs of "my Vector is cylindrical, but my Transformation is spherical"). Transformations, which mix coordinates, should all be stored as Cartesian for simplicity. Vectors can store data in different coordinate systems, but they get converted to Cartesian when being transformed, just as some operations, like `+`, should collapse to Cartesian.

**Reply by nsmith-** [reply]
I think that's fair that there will be a set of operations (transforms, addition, perhaps others) that necessarily convert to cartesian storage. One nice feature might be a way for users to convert the storage back to spherical or otherwise, if it is known to be more efficient for subsequent processing.

**Comment by nsmith-** [comment]
> Base arrays/objects have purely geometric names for their arguments, such as perp instead of pt, but overloads mix in synonyms.

If we are to make a fully separate set of positional and momentum property/method names, I think it would be nice to be able to optionally expose only one or the other.

**Reply by jpivarski** [reply]
It's hard to remove properties from an object, and doing so would involve some Python magic that probably won't play well with typing. It would be possible to get this by creating a momentum class that isn't a subclass of the geometric class, passing each member through, but then a lot of infrastructure would have to be duplicated for that because a lot of this works by inheritance (mix-ins).

I wonder, though, how important that would be. If you had a momentum vector or an array of momentum vectors with a `pt` property, is it a problem that it also has a `perp` property? After all, the geometric concepts are still valid when a vector is interpreted as momentum, but the reverse is not true. If you're using a vector for secondary vertex displacement, `pt` is potentially misleading, but `perp` is right, whereas if you're using a vector for momentum, `pt` and `perp` are both right.

**Reply by nsmith-** [reply]
One argument is the proliferation of attributes, having `x, y, z, px, py, pz` can get a bit crowded, especially if other record properties are comingled. But that is just aesthetics. Some objects, such as secondary vertices, have an associated position and momentum vector. There one might want to expose one or the other separately (though perhaps the answer there is to use a subrecord for each)

**Reply by jpivarski** [reply]
A secondary vertex (position in the detector) and the momentum sum of the particles produced at that vertex (momentum direction) really, really should be subrecords. Not just for this reason.

And anyway, ROOT Lorentz vectors have every synonym under the sun for every vector. By not including momentum-related words on non-momentum vectors and including pure geometric words on every vector, that is some clean-up. Arguably, we shouldn't clean up any more than that, because a momentum vector "is a" direction in a space, just not position space.

**Reply by henryiii** [reply]
I'm also perfectly fine with a minimal, basis set of properties based on pure geometric words; adding an interpretation doesn't need to invalidate the basis (it can't in Python, of course, but I don't think it even really ideally needs to).

**Comment by henryiii** [comment]
Let's avoid list 2 point 8 for now: as I've pointed out many times, I am _strongly_ against using 3D vectors for positions; they are _not_ points, they are directions + magnitude. The first iteration should not give any support to affine vectors at all. Functions like `point + point` are actually well defined - but not if you use a 3D vector for a position; because 3D position can't be mathematically represented with three components; only direction can! There will be two eventual possibilities; either someone can extend this for affine vectors in a standalone package (and they can take all the bugs associated with trying to do this with three coordinates), or we can eventually add a real 4D Cartesian vector and it's representation as a 3D point in space. There are well defined methods from computer graphics / ray tracing for working with points in this way.

**Reply by jpivarski** [reply]
If we don't want vectors to be misused as points (I agree: it was very useful to distinguish them in alignment), then providing only vectors will encourage just that. At least we need to include the names so that it's a concept that can be filled in—just leaving it unspecified invites abuse.

**Reply by nsmith-** [reply]
Can you put some references here for the 4-component 3D point? I remember being thoroughly convinced by you once but none of the details :)

**Reply by jpivarski** [reply]
A coordinate frame transformation (translation + rotation in 3D, for example) can be accomplished by adding a 3D vector and multiplying a 3×3 matrix (or in the other order, but pick a convention). This is the same as having a 4D vector in which the last row and column encodes the 3D displacement vector. That is, `A·x + b` with `A` being 3×3 and `b` being a vector like `x` can be replaced with `B·y` where `B` is 4×4 and `y` is 4-dimensional with the last component being `1`.

If you have a lot of machinery for computing matrix multiplications already, this is a neat trick that simplifies code—it's all matrix multiplication, no need for that extra (manual) addition step. However, that doesn't help us much here, since we're concerned with scaling in the other direction: functions that will generically run through NumPy, Awkward, Numba, etc. would have to pick out components by name, such as `xx`, `xy`, `xz`, `yy`, `yz`, `zz`, as we can't assume that `xx` is located next to `xy` in memory, as a matrix library generally would. But all of our `xx` values are equal strides away from other `xx` values, so we'll be vectorizing that matrix multiplication, term by term. Since we're writing out the matrix multiplication manually, the slickness of turning a 3×3 matrix into a 4×4 matrix to remove the extra `+` doesn't apply. On the other hand, anyone who doesn't know the 3×3 → 4×4 trick will have an easier time reading our code.

**Reply by nsmith-** [reply]
..which allows to compose transformations without reference to a particular point, nice

**Reply by henryiii** [reply]
If you look at additions, you'll realize the last component is actually the "weight". Adding two vectors with equal weight gives you the midpoint between them. The extra component is required to turn a direction into a position - a direction is tied to a special point - we generally make that point the 'origin' for positions. Adding an extra DoF removes the dependence on that special point, which is why things like additions and transformations suddenly make sense again without the origin affecting things.

**Reply by henryiii** [reply]
(You divide by the final point when going from the 4D vector back to the three position coordinates; so if the weight is 2, you divide the other values by 2 to scale it back to a normal, weight 1 point so that you can read the position coordinates directly off).

**Reply by jpivarski** [reply]
Oh, I get it: I've only ever seen that extra component as `1`, and I haven't assumed that `point + point` would be meaningful. But I suppose interpreting the component as a weight would make `point + point` be the average, optionally weighted.

I'm not sure that I want to get into all of that right away, but at least I want to have a place for it. I'm working on a wiki now to describe an ontology and give names to things. Operations between vectors and vectors result in other vectors or scalars, but they do not result in points (the other direction is true: `point - point` is a vector). Thus, we don't have to implement points right away, but it helps to recognize that they exist, so that the naming conventions are capable of including them.

**Reply by henryiii** [reply]
The center of mass of a collection of weighed points is ∑p :)

**Reply by jpivarski** [reply]
I encountered the following and thought of this thread:

> You can add two vectors but not two points and the difference between two points is a vector.

From the [ROOT GenVector documentation](https://root.cern.ch/doc/master/group__GenVector.html).

It all depends on what you admit into your mathematical system. I don't think we're al very interested in weighted averages of points, though I think it's pretty neat that it naturally comes out of the 4D representation of 3D transformations. (We should make it possible to turn points into vectors and vice-versa, but make the point interface rather limited to prevent accidental misuse. As in the Python philosophy, all misuse should be intentional!)

**Reply by eduardo-rodrigues** [reply]
Hi @jpivarski, all: on this point let me divert you to what Vanya and I had added in the old times to the scikit-hep package for lines and planes in 3D - https://github.com/scikit-hep/scikit-hep/blob/master/skhep/math/geometry.py. Not sure how easy it is to take that functionality into this very more powerful "framework".

**Reply by jpivarski** [reply]
Thanks for the reminder! I've been looking at the Scikit-HEP geometry packages—they're denoted with an S in a circle on the right side of this page (checking all the properties and methods of earlier vector packages, to make sure I haven't missed one; on the left is a set of collated names with synonyms).

![IMG_20210308_082808108](https://user-images.githubusercontent.com/1852447/110334535-7f4fab80-7fe8-11eb-9ec1-63d4cbcf590f.jpg)

**Reply by henryiii** [reply]
I would hope the Geometry package would become a standalone package dependent on Vector. Like GeometryVector or something similar.

**Reply by eduardo-rodrigues** [reply]
That's a possibility for sure. I just thought I would point the code out as it seems relevant to the discussions and design ongoing. You know from previous chats that I have no strong preference and would be totally happy to see line and like also here, though I see why you make the comment.

**Reply by eduardo-rodrigues** [reply]
BTW, @jpivarski, great notebook 👍. I still work with paper a lot myself - nothing like it!

**Reply by jpivarski** [reply]
I'm not planning on doing lines and stuff (math/geometry.py as opposed to math/vector.py). Actually, the scope is pretty well encapsulated by the above page.

**Comment by jpivarski** [comment]
I'm using the wiki to [write up ontologies and naming conventions](https://github.com/scikit-hep/vector/wiki). This is mainly so that I can give names to a large number of `compute` functions and later will also be helpful in setting up the class hierarchy with appropriate mix-ins.

**Comment by jonas-eschle** [comment]
Great to see this and I'll gladly help with the implementation of TensorFlow and friends. I think they (jax, TF, PyTorch) share a commonality: they all have a numpy API, but don't allow in-place assignments (AFAIK), which we may want for e.g. pure Numpy.

**Reply by jpivarski** [reply]
Awkward Array doesn't allow in-place assignments, either, but when you're working with NumPy, you expect that. Therefore, the general classes won't assume but won't prevent mutability—that will be implemented independently on the backends that support it. That's how I was thinking about supporting it.

We should also keep an eye on the [Consortium for Python Data API Standards](https://data-apis.org/blog/announcing_the_consortium/), which is standardizing NumPy-like APIs (with the same view toward optional mutability). What they have in mind isn't general enough to include Awkward Array, but it would include all the backends, and maybe we can get a lot of them at once through some protocol this group develops.

**Comment by jpivarski** [comment]
A follow-up: the bulk implementation of Vector is pretty much done. [I've written an overview here](https://nbviewer.jupyter.org/github/scikit-hep/vector/blob/develop/docs/usage/intro.ipynb), which shows how the whole project fits together. Tomorrow, I'm writing docstrings.

At this point, we should start encouraging adventurous users to use it. Doubtless, there will be rough edges that need to be fixed—I found a few of them while writing the overview. I'll be pivoting to work on Uproot in April, and one of the things I ought to do is make Uproot present `TLorentzVector`, `ROOT::Math::LorentzVector`, and similar as Vector's object and array types.

Oh, we'd also need to transition the `develop` branch to `main` and publish a release. I wonder if @henryiii wants to look it over before taking that step.

Take a look!

**Reply by henryiii** [reply]
Yes, maybe we can aim for a release on Wednesday or so? I'd like to go over it, touch up a few things (like making sure the docs are showing up on RtD, currently untested, I think, due to the branch changes, making sure `py.typed` is present in the output wheel, etc).

**Reply by eduardo-rodrigues** [reply]
I'm sure to be having a look. From a quick glance it looks beautiful!

**Reply by jpivarski** [reply]
> making sure the docs are showing up on RtD

This is missing: the .rst stubs will have to be recreated. I deleted the original ones because the module names are different now, but the new names (modules, classes, methods) are stable—there won't be any new ones before Wednesday.

Given the nature of what these functions do, I'll want to include some math formulas in the docstrings. I know that reST can do math, but it's horrible to read in text form. (I really don't like reST...) I'm going to see if there's something that will convert simple TeX between `$` or `$$` or something into the appropriate output for RtD, maybe a Sphinx extension.

**Reply by henryiii** [reply]
https://www.sympy.org/sphinx-math-dollar/
